node/core/approval-voting/src/aux_schema/mod.rs

```diff
@@ -0,0 +1,531 @@
+		let mut pruned_branches = Vec::new();
+
+		for b in at_height {
+			let children = visit_and_remove_block_entry(
+				b,
+				&mut deleted_block_keys,
+				&mut visited_candidates,
+			)?;
+
+			if b != canon_hash {
+				pruned_branches.extend(children);
+			}
+		}
+
+		pruned_branches
+	};
+
+	// Follow all children of non-canonicalized blocks.
+	{
+		let mut frontier: Vec<_> = pruned_branches.into_iter().map(|h| (canon_number + 1, h)).collect();
+		while let Some((height, next_child)) = frontier.pop() {
+			let children = visit_and_remove_block_entry(
+				next_child,
+				&mut deleted_block_keys,
+				&mut visited_candidates,
+			)?;
+
+			// extend the frontier of branches to include the given height.
+			frontier.extend(children.into_iter().map(|h| (height + 1, h)));
+
+			// visit the at-height key for this deleted block's height.
+			let at_height = match visited_heights.entry(height) {
+				Entry::Occupied(e) => e.into_mut(),
+				Entry::Vacant(e) => e.insert(load_blocks_at_height(store, height)?),
+			};
+
+			if let Some(i) = at_height.iter().position(|x| x == &next_child) {
+				at_height.remove(i);
+			}
+		}
+	}
+
+	// Update all `CandidateEntry`s, deleting all those which now have empty `block_assignments`.
+	let (written_candidates, deleted_candidates) = {
+		let mut written = Vec::new();
+		let mut deleted = Vec::new();
+
+		for (candidate_hash, candidate) in visited_candidates {
+			if candidate.block_assignments.is_empty() {
+				deleted.push(candidate_entry_key(&candidate_hash));
+			} else {
+				written.push((candidate_entry_key(&candidate_hash), candidate.encode()));
+			}
+		}
+
+		(written, deleted)
+	};
+
+	// Update all blocks-at-height keys, deleting all those which now have empty `block_assignments`.
+	let written_at_height = {
+		visited_heights.into_iter().filter_map(|(h, at)| {
+			if at.is_empty() {
+				deleted_height_keys.push(blocks_at_height_key(h));
+				None
+			} else {
+				Some((blocks_at_height_key(h), at.encode()))
+			}
+		}).collect::<Vec<_>>()
+	};
+
+	// due to the fork pruning, this range actually might go too far above where our actual highest block is,
+	// if a relatively short fork is canonicalized.
+	let new_range = StoredBlockRange(
+		canon_number + 1,
+		std::cmp::max(range.1, canon_number + 2),
+	).encode();
+
+	// Because aux-store requires &&[u8], we have to collect.
+
+	let inserted_keys: Vec<_> = std::iter::once((&STORED_BLOCKS_KEY[..], &new_range[..]))
+		.chain(written_candidates.iter().map(|&(ref k, ref v)| (&k[..], &v[..])))
+		.chain(written_at_height.iter().map(|&(ref k, ref v)| (&k[..], &v[..])))
+		.collect();
+
+	let deleted_keys: Vec<_> = deleted_block_keys.iter().map(|k| &k[..])
+		.chain(deleted_height_keys.iter().map(|k| &k[..]))
+		.chain(deleted_candidates.iter().map(|k| &k[..]))
+		.collect();
+
+	// Update the values on-disk.
+	store.insert_aux(
+		inserted_keys.iter(),
+		deleted_keys.iter(),
+	)?;
```

Comment:
`Vec<_>` implements `IntoIterator` already.

```suggestion
	store.insert_aux(inserted_keys, deleted_keys)?;
```

Reply:
Ah, I see. Well, I certainly agree that `insert_aux`'s type signature could be more ergonomic!